engine/src/conversion/analysis/fun/implicit_constructors.rs

```diff
@@ -59,261 +77,481 @@ struct ExplicitFound {
-                })
-            });
-            // Conservatively, we will not generate implicit constructors for any struct/class
-            // where we don't fully understand all field types. We need to extend our knowledge
-            // to understand the constructor behavior of things in known_types.rs, then we'll
-            // be able to cope with types which contain strings, unique_ptrs etc.
-            let any_field_or_base_not_understood = bases
-                .iter()
-                .chain(field_types.iter())
-                .any(|qn| !all_known_types.contains(qn));
-            let explicit_items_found = ExplicitItemsFound {
-                move_constructor: find(ExplicitKind::MoveConstructor),
-                copy_constructor: find(ExplicitKind::ConstCopyConstructor)
-                    || find(ExplicitKind::NonConstCopyConstructor)
-                    || find(ExplicitKind::DeletedOrInaccessibleCopyConstructor),
-                any_other_constructor: find(ExplicitKind::OtherConstructor),
-                any_bases_or_fields_lack_const_copy_constructors,
-                any_bases_or_fields_have_deleted_or_inaccessible_copy_constructors,
-                any_bases_have_deleted_or_inaccessible_destructors,
-                destructor: find(ExplicitKind::Destructor)
-                    || find(ExplicitKind::DeletedOrInaccessibleDestructor),
-                copy_assignment_operator: find(ExplicitKind::CopyAssignmentOperator),
-                move_assignment_operator: find(ExplicitKind::MoveAssignmentOperator),
-                has_rvalue_reference_fields: details.has_rvalue_reference_fields,
-                any_field_or_base_not_understood,
+            let get_items_found = |qn: &QualifiedName| -> Option<ItemsFound> {
+                if known_types().is_known(qn) {
+                    Some(known_type_items_found(qn))
+                } else {
+                    all_items_found.get(qn).copied()
+                }
             };
-            log::info!(
-                "Explicit items found for {:?}: {:?}",
-                name,
-                explicit_items_found
-            );
-            let implicits = determine_implicit_constructors(explicit_items_found);
-            implicit_constructors_needed.insert(name.clone(), implicits);
+            let bases_items_found: Vec<_> = bases.iter().map_while(get_items_found).collect();
+            let fields_items_found: Vec<_> = field_info
+                .iter()
+                .filter_map(|field_info| match field_info.type_kind {
+                    TypeKind::Regular | TypeKind::SubclassHolder(_) => match field_info.ty {
+                        Type::Path(ref qn) => get_items_found(&QualifiedName::from_type_path(qn)),
+                        _ => None,
+                    },
+                    // TODO: Figure out how to differentiate between pointers and references coming
+                    // from C++. Pointers have a default constructor.
```

Comment:
Could you file a bug for this and ideally a test case? I'm not totally clear on which bits of this aren't done.

Reply:
Opened #865, and added comments on the parts of the test being added in this change which demonstrates it.